engine/common/synchronization/finalized_snapshot.go

```diff
@@ -0,0 +1,139 @@
+type finalizedSnapshot struct {
+	head         *flow.Header
+	participants flow.IdentityList
+}
+
+// FinalizedSnapshotCache represents a cached snapshot of the latest finalized header and participants list.
+// It is used in Engine to access latest valid data.
+type FinalizedSnapshotCache struct {
+	mu sync.Mutex
+
+	log                       zerolog.Logger
+	state                     protocol.State
+	identityFilter            flow.IdentityFilter
+	lastFinalizedSnapshot     *finalizedSnapshot
+	finalizationEventNotifier engine.Notifier // notifier for finalization events
+
+	lm                *lifecycle.LifecycleManager
+	shutdownCompleted chan struct{} // used to signal that shutdown of finalization processing loop was completed
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewFinalizedSnapshotCache creates a new finalized snapshot cache.
+func NewFinalizedSnapshotCache(log zerolog.Logger, state protocol.State, participantsFilter flow.IdentityFilter) (*FinalizedSnapshotCache, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	cache := &FinalizedSnapshotCache{
+		state:             state,
+		identityFilter:    participantsFilter,
+		lm:                lifecycle.NewLifecycleManager(),
+		ctx:               ctx,
+		cancel:            cancel,
+		shutdownCompleted: make(chan struct{}),
+		log:               log.With().Str("component", "finalized_snapshot_cache").Logger(),
+	}
+
+	err := cache.updateSnapshot()
+	if err != nil {
+		return nil, fmt.Errorf("could not apply last finalized state")
+	}
+
+	return cache, nil
+}
+
+// get returns last locally stored snapshot which contains final header
+// and list of filtered identities
+func (f *FinalizedSnapshotCache) get() *finalizedSnapshot {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	return f.lastFinalizedSnapshot
+}
+
+// updateSnapshot updates latest locally cached finalized snapshot
+func (f *FinalizedSnapshotCache) updateSnapshot() error {
+	finalSnapshot := f.state.Final()
+	head, err := finalSnapshot.Head()
+	if err != nil {
+		return fmt.Errorf("could not get last finalized header: %w", err)
+	}
+
+	// get all participant nodes from the state
+	participants, err := finalSnapshot.Identities(f.identityFilter)
+	if err != nil {
+		return fmt.Errorf("could get consensus participants at latest finalized block: %w", err)
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.lastFinalizedSnapshot != nil && f.lastFinalizedSnapshot.head.Height >= head.Height {
+		return nil
+	}
+
+	f.lastFinalizedSnapshot = &finalizedSnapshot{
+		head:         head,
+		participants: participants,
+	}
+
+	return nil
+}
+
+func (f *FinalizedSnapshotCache) Ready() <-chan struct{} {
+	f.lm.OnStart(func() {
+		go f.finalizationProcessingLoop(f.ctx)
+	})
+	return f.lm.Started()
+}
+
+func (f *FinalizedSnapshotCache) Done() <-chan struct{} {
+	f.cancel()
+	f.lm.OnStop(func() {
+		// wait for finalization processing loop to shutdown
+		<-f.shutdownCompleted
```

Comment:
Same question here, using `unit.Launch` makes things simpler

Reply:
I am deliberately not using `engine.unit` here, for multiple reasons:
* it does not guarantee idempotency of `Ready` or `Done`, nor concurrency safety. `LifecycleManager` [does](https://github.com/onflow/flow-go/pull/1031)
* it was originally intended for use by engines, which this is not. we are not really taking advantage of any of the useful features of it here

The first point is the more important one.

Reply:
Can we have same useful interface for `LifecycleManager` as we have for unit? `unit.Launch` is really easy to use in terms of waiting for completion and canceling, seems like `LifecycleManager` can be combined with unit to provide concurrency safety + useful and easy to use API

Reply:
It appears we've both arrived at the same conclusion :) https://github.com/dapperlabs/flow-go/issues/5720

`unit` does implement some antipatterns that I would like to get rid of, but I agree with your point about the useful functionality it provides.